README.md

```diff
@@ -33,106 +33,131 @@ A project was created with those packages `npm install`ed, creating a gigantic `
 
-#### `import` statements of ESM packages
+### Parsing files as ESM or as CommonJS
 
-Importing of ESM packages is covered by the [package exports proposal](https://github.com/jkrems/proposal-pkg-exports).
+There are four types of `import` statement specifiers:
 
-#### `import` statements of ESM files
+1. “Bare” specifiers like `'lodash'`, which are the name of a package and refer to the package entry point.
+2. “Deep import” specifiers like `'lodash/lib/shuffle.mjs'`, which refer to a file within a package.
+3. Relative file specifiers like `'./startup.js'` or `'../config.mjs'`, which refer to a file relative to the importing file.
+3. Absolute URL file specifiers like `'file:///opt/nodejs/config.js'`, which refer directly to a file.
 
-An ESM JavaScript file would use an `import` statement to import another ESM JavaScript file. Imported files must have either a `.js` or `.mjs` extension, and must be specified using a relative path or absolute URL. (Absolute paths may be supported in the future, if a system can be worked out that is compatible with browsers; see also Absolute URLs below.)
+In all cases, first Node follows its algorithm to locate a file to load. Once the file is found, Node must then decide whether to load it as ESM or as CommonJS. The algorithm goes as follows:
 
-Paths must begin with `./` or `../`; specifiers that begin with an allowed package name character are treated as bare specifiers/package names and are covered by the [package exports proposal](https://github.com/jkrems/proposal-pkg-exports). Like import specifiers in browsers, Node’s import specifiers are URLs and therefore always use forward slashes and escape characters the same way URLs do.
-
-##### Relative to the importing file: specifiers starting with `./` or `../`
-
-The following `import` statements import relative files, treating them as JavaScript with an ESM parse goal:
-
-```js
-// Folder structure:
-// - index.js (this file)
-// - package.json
-// - constants.mjs
-// - helpers/temperature.mjs
-
-import { APP_ROOT_URL } from './constants.mjs';
-
-import { convertTemperature } from './helpers/temperature.mjs';
 ```
+If the file is a package entry point
+    And the package’s package.json is detected to be ESM
+        Load the file as ESM.
+    Else
+        Load the file as CommonJS.
+Else
+    If there is a package.json in the folder where the file is
+        And the package.json is detected to be ESM
+            Load the file as ESM.
+        Else
+            Load the file as CommonJS.
+    Else
+        Go into the parent folder and look for a package.json there
+        Repeat until we either find a package.json or hit the file system root
+        If we found a package.json
+            And the package.json is detected to be ESM
+                Load the file as ESM.
+            Else
+                Load the file as CommonJS.
+        Else we reach the file system root without finding a package.json
+            Load the file as ESM.
```

Comment:
Above you clearly state: **There are (at least) two parts to module resolution: _location_ and _interpretation_.**, which I completely agree with.

Then in this spec you conflate the two.

Perhaps lets maintain that separation in the specs entirely. While the package exports proposal does need to check the package boundary / format, that could still be a separate process (and a shared cache in implementations).

So I was thinking of something more like:

```js
**GET_MODULE_FORMAT(url)**:
1. Let _pjson_ be READ_PACKAGE_BOUNDARY(url)
1. If _pjson_ is _undefined_ or has an _exports_ or _mode_ then,
  1. Return "esm"
1. Return "cjs"
```

Reply:
Yes, we need to update the resolution spec to cover all this, assuming you like this proposal.

Where am I conflating the two? I was trying to avoid discussing location other than when necessary, like to explain the concept of package scopes. The stuff about entry points and dual-mode packages should arguably go in the package exports proposal, or we could merge the two.

Reply:
@guybedford can you clarify if any the changes I would want to make here please.

Reply:
@GeoffreyBooth I guess my point is that the short spec I wrote above is exactly equivalent to what you have written.

Reply:
So do you think including it here is relevant and if so does it replace something else?

cc/ @GeoffreyBooth

Reply:
Sorry @guybedford… okay, so I think I skipping a line in head scrolling through this. Basically, this replaces this entire fenced block 😊

Reply:
@guybedford So re locating vs interpreting, I'm skipping the locating of the file to be imported, e.g. foo.js, but I'm describing the search for package.json and therefore the package scope. I think the latter is important to include here, as it's a new concept and really the central concept to this proposal. I think it should be spelled out in more detail and more plain language than the spec style you wrote above, to make sure people understand; but the spec snippet you wrote here should certainly go in the resolution spec document, or we could include it here in addition to the plain language psuedocode.

Reply:
Ahh I didn't reread the spec properly. What you have now is effectively the READ_PACKAGE_BOUNDARY function I was pointing to there. Looks good.

Reply:
@GeoffreyBooth okay, this makes more sense now. But let me make sure I got it:

1. **in scope** Given a non-absolute specifier, convert to absolute URL.
2. **in scope** Starting from the parent folder for the absolute URL of the imported module:
  a. Request to check for and load `package.json` from current folder. (actual loading out of scope)
  b. if found and parsed, return the parsed data and the path to the `package.json`.
  c. if found and not loaded, return with error details for the invalid `package.json`
  d. If not found:
    1. If no parent folder exists, return with "no package scope" (not sure)
    2. Else move to parent folder and continue from 2a

Reply:
@GeoffreyBooth if you have time, I want to make sure I got it right before I make a mess 😄